File: doubly_linked_list.py
class Node:
    def __init__(self,data):
        self.data = data
        self.left = None
        self.right = None
class header:
    def __init__(self):
        self.head = None
        self.tail = None
class operations:
    def insAtBeg(self,d,length):
        temp = Node(d)
        hdr = h.head
        if(hdr == None):
            h.head = temp
        else:
            temp.right = h.head
            hdr.left = temp
            h.head = temp
        if(length == 1):
            h.tail = temp
    def insAtEnd(self,d,length):
        temp = Node(d)
        h.tail.right = temp
        h.tail = temp
        
    def insAtPos(self,pos,d,length):
        temp = Node(d)
        if(pos == 1):
            self.insAtBeg(d,length)
        else:
            c= 2
            trav = h.head
            while(c != pos):
               c+=1
               trav = trav.right
            temp.right = trav.right
            x = temp.right
            x.left = temp
            trav.right = temp
            temp.left = trav
    def delAtBeg(self,length):
        temp = h.head
        h.head = temp.right
        if(h.head !=None):
            h.head.left = None
        del temp
        if(length == 1):
            h.tail = temp
    def delAtEnd(self,length):
        temp = h.tail
        h.tail = h.tail.left
        tail.right = None
        del temp
    def delAtPos(self,pos,length):
        if(pos == 1):
            self.delAtBeg(length)
        else:
            c= 2
            trav = h.head
            while(c!=pos):
                c+=1
                trav = trav.right
            temp = trav.right
            trav.right = temp.right
            x = trav.right
            if(x != None):
                x.left = trav
            else:
                h.tail = trav
            del temp
    def search(self,pos,length):
        if(pos> length//2):
            trav = h.tail
            c = length-pos
            while(c>0):
                c-=1
                trav = trav.left
            print(trav.data)
        else:
            c= 1
            trav = h.head
            while(c != pos):
                trav = trav.right
            print(trav.data)
    def print(self):
        trav = h.head
        while(trav != None):
            print(trav.data,end =" ")
            trav = trav.right
            
                



h = header()
op = operations()
length = 0
print("Linked list implementation")
choice = int(input("1. Insert at beginning \n2.Insert at end\n3.Insert at pos\n4.Deletion at beginning\n5.Deletion at end\n6.Deletion at position\n7.Search\n8.Print elements\n9.Exit\n"))
while(choice != 9):
    if(choice == 1):
        data = int(input("Enter the data to enter"))
        length+=1
        op.insAtBeg(data,length)
    elif(choice == 2):
        data = int(input("Enter the data to enter"))
        length+=1
        op.insAtEnd(data,length)
    elif(choice == 3):
        data = int(input("Enter the data to enter"))
        pos =  int(input("Enter the pos to enter"))
        length+=1
        op.insAtPos(pos,data,length)
    elif(choice == 4):
        length-=1
        op.delAtBeg(length)
    elif(choice == 5):
        length-=1
        op.delAtEnd(length)
    elif(choice == 6):
        length-=1
        pos = int(input("Enter the pos to delete"))
        op.delAtPos(pos,length)
    elif(choice == 7):
        pos = int(input("Enter the pos to search"))
        op.search(pos,length)
    elif(choice == 8):
        op.print()
    print(length,"is the length of the list")
    choice = int(input("Enter your choice"))

File: doubly_linked_list.java
package doublylinkedlist;

import java.util.Scanner;
public class Node {
    int data;
    Node right;
    Node left;
    Node(int d)
    {
        data = d;
        left = null;
        right = null;
    }
}


public class DoublyLinkedList {
     Node head;
    Node tail;
    
    private void InsertAtBeg(int data) {
        Node node = new Node(data);
        if(head == null)
        {
            head = node;
            tail = node;
        }
        else
        {
         Node temp = head;
         node.right = temp;
         temp.left = node;
         head = node;
        }
    }
    private void InsertAtEnd(int data) {
        if(head == null){
         InsertAtBeg(data);   
        }
        else
        {
        Node node = new Node(data);
        node.left = tail;
        tail.right = node;
        tail = node;
        }
    }
    private void InsertAtPos(int data, int pos, int count) {
        if(pos == 1)
            InsertAtBeg(data);
        else if(pos == count+1)
            InsertAtEnd(data);
        else
        {
        Node node = new Node(data);
        int posfind = 2;
        if(count/2 > pos)
        {
         Node trav = head;
         while(posfind < pos-1)
         {
             trav = trav.right;
             posfind++;
         }
         Node temp = trav.right;
         trav.right = node;
         node.right = temp;
         temp.left = node;
         node.left = trav;
        }
        else
        {
            pos = count-pos;
            posfind = 1;
            Node trav = tail;
            while(posfind < pos-1)
            {
               trav = trav.left;
               posfind++;
            }
            Node temp = trav.left;
            trav.left = node;
            node.left = temp;
            temp.right = node;
            node.right = trav;
        }
        }
    }
    private void DeleteAtBeg() {
        Node temp = head;
        head = temp.right;
        head.left = null;
        
    }
    private void DeleteAtEnd() {
        Node temp = tail;
        tail = temp.left;
        tail.right = null;
    }
    private void DeleteAtPos(int pos, int count) {
        if(pos == 1)
        {
            DeleteAtBeg();
        }
        else if(pos == count)
        {
            DeleteAtEnd();
        }
        else if(pos< count/2)
        {
            int posfind = 1;
            Node trav = head;
            while(posfind <pos-1)
            {
                trav = trav.right;
                posfind++;
            }
            Node temp = trav.right;
            trav.right = temp.right;
            temp.right.left = trav;
            temp.right = null;
            temp.left = null;
        }
        else {
            pos = count-pos;
            int posfind = 1;
            Node trav = tail;
            while(posfind<pos-1)
            {
                trav = trav.left;
                posfind++;
            }
            Node temp = trav.left;
            trav.left = temp.left;
            temp.left.right = trav;
            temp.right = null;
            temp.left = null;
            
        }
    }
    private int Search(int data) {
        int res = -1;
        int posfind = 1;
        Node trav = head;
        while(trav != null)
        {
            if(trav.data == data)
            {
                res = posfind;
            }
            trav = trav.right;
            posfind++;
        }
        return res;
    }
    private void display() {
        Node trav = head;
        while(trav != null)
        {
            System.out.println(trav.data);
            trav = trav.right;
        }
    }

    
    public static void main(String[] args) {
        System.out.println("Doubly linked list\n");
        Scanner scan = new Scanner(System.in);
        DoublyLinkedList dll = new DoublyLinkedList();
        int count = 0;
        int data,pos;
        System.out.print("Enter your choice\n 1.Insert at beginning\n2.Insert at end\n3.Insert at pos\n4.Delete at beginning\n5.Delete at end\n6.Delete at pos\n7.Search\n8.Display\n9.Exit");
        int choice = scan.nextInt();
        while(choice != 9)
        {
            switch(choice)
            {
                case 1:
                    System.out.println("Enter the data to insert");
                    data = scan.nextInt();
                    dll.InsertAtBeg(data);
                    count++;
                    dll.display();
                    break;
                case 2:
                    System.out.println("Enter the data to insert");
                    data = scan.nextInt();
                    dll.InsertAtEnd(data);
                    count++;
                    dll.display();
                    break;
                case 3:
                    System.out.println("Enter the data to insert");
                    data = scan.nextInt();
                    System.out.println("Enter the position to insert");
                    pos = scan.nextInt();
                    dll.InsertAtPos(data,pos,count);
                    count++;
                    dll.display();
                    break;
                case 4:
                    dll.DeleteAtBeg();
                    count--;
                    dll.display();
                    break;
                case 5:
                    dll.DeleteAtEnd();
                    count--;
                    dll.display();
                    break;
                case 6:
                    System.out.println("Enter the position to delete");
                    pos = scan.nextInt();
                    dll.DeleteAtPos(pos,count);
                    count--;
                    dll.display();
                    break;
                case 7:
                    System.out.println("Enter the element to search");
                    data = scan.nextInt();
                    pos = dll.Search(data);
                    if(pos == -1)
                        System.out.println("The eleement is not in the list");
                    else
                        System.out.println("The eleement is presenet in position "+pos);
                    break;
                case 8:
                    dll.display();
                    break;               
            }
            System.out.println("Enter your choice of number");
            choice = scan.nextInt();
        }
    }
    
}
